### PR TITLE
BUG: Python Parser skipping over items if BOM present in first element of header

### LINKED ISSUE #36343: BUG: ZERO WIDTH NO-BREAK SPACE in column name causes a reading failure
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

[test excel.xlsx](https://github.com/pandas-dev/pandas/files/5214984/test.excel.xlsx)

```python
# Your code here
import pandas as pd
df=pd.read_excel('test excel.xlsx')
df

```

#### Problem description

I have a customer that is sending me a file that has ZERO WIDTH NO-BREAK SPACE
only the first header reads unless the character is deleted 
I created a repro by copying the header from the original file to the new one in excel

#### Expected Output

``` 
Building Park Name	Building Name	New / Renewal	Address	City	State	Class 
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.6.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
machine          : AMD64
processor        : Intel64 Family 6 Model 158 Stepping 10, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : None.None

pandas           : 0.25.3
numpy            : 1.18.1
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 20.1.1
setuptools       : 41.2.0
Cython           : None
pytest           : 5.3.5
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : 1.2.8
lxml.etree       : None
html5lib         : 1.0.1
pymysql          : None
psycopg2         : None
jinja2           : 2.10.3
IPython          : 7.11.1
pandas_datareader: None
bs4              : 4.8.2
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : 3.1.2
numexpr          : None
odfpy            : None
openpyxl         : 3.0.3
pandas_gbq       : None
pyarrow          : None
pytables         : None
s3fs             : None
scipy            : 1.4.1
sqlalchemy       : 1.3.12
tables           : None
xarray           : None
xlrd             : 1.2.0
xlwt             : None
xlsxwriter       : 1.2.8
</details>

### PR DESCRIPTION
**Pull request by asishm**
- [x] closes #36343
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/io/parsers.py [hunk 1]
```diff
@@ -2876,7 +2876,7 @@ def _check_for_bom(self, first_row):
             return [new_row] + first_row[1:]
 
         elif len(first_row_bom) > 1:
-            return [first_row_bom[1:]]
+            return [first_row_bom[1:]] + first_row[1:]
```
Comment (WillAyd) [inline pandas/io/parsers.py:2879]: Hmm I don't find this very clear why we would do this - can you try refactoring the code above this to better suit the requirement?

**Review by WillAyd — CHANGES_REQUESTED** [review]
Thanks for taking a look. Can you also add a whatsnew note for 1.2?

File: pandas/io/parsers.py (hunk 1 above)
Comment (asishm) [reply pandas/io/parsers.py]: from what I understand
the code block above removes the BOM from the first element of the row and removes the quotes if it was a quoted string. i.e. if it was of the format `<BOM><QUOTE>abc<QUOTE>def` then it extracts out `abcdef`

I've tried to refactor it slightly

File: pandas/tests/io/parser/test_common.py [hunk 2]
```diff
@@ -2127,6 +2127,12 @@ def test_first_row_bom(all_parsers):
     expected = DataFrame(columns=["Head1", "Head2", "Head3"])
     tm.assert_frame_equal(result, expected)
 
+    # see gh-36343
```
Comment (jreback) [inline pandas/tests/io/parser/test_common.py:2130]: can you make this a new test

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
@gfyoung if you'd have a look

**Comment by gfyoung** [comment]
@jreback's comments notwithstanding, this looks good.  Nice catch @asishm !

File: pandas/tests/io/parser/test_common.py (hunk 2 above)
Comment (asishm) [reply pandas/tests/io/parser/test_common.py:2132]: done, I think test failures are unrelated. https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=42772&view=ms.vss-test-web.build-test-results-tab&runId=2173848&resultId=150762&paneView=debug

**Review by WillAyd — APPROVED** [review]
lgtm when green

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @asishm 
nice. I *think* we might have some other BOM/EOM issues in the tracker if you'd have a look (and maybe this will close them).

**Comment by asishm** [comment]
@jreback 

I searched for 'bom' and only found #31609 which can be closed - it talks about the same solution as this PR. not really sure what else to search for. (eom doesn't lead to anything)

**Comment by jreback** [comment]
k thanks @asishm yeah there might be some others but IIRC those are for the c-parser

**Pull request merged**